appx.c

```diff
@@ -0,0 +1,2645 @@
+}
+
+void bioAddU16(BIO *bio, uint16_t v)
+{
+	uint8_t b[2];
+
+	b[0] = v & 0xFF;
+	b[1] = (v >> 8) & 0xFF;
+
+	BIO_write(bio, b, 2);
+}
+
+void bioAddU32(BIO *bio, uint32_t v)
+{
+	uint8_t b[4];
+
+	b[0] = v & 0xFF;
+	b[1] = (v >> 8) & 0xFF;
+	b[2] = (v >> 16) & 0xFF;
+	b[3] = (v >> 24) & 0xFF;
+
+	BIO_write(bio, b, 4);
+}
+
+void bioAddU64(BIO *bio, uint64_t v)
+{
+	uint32_t l = v & 0xFFFFFFFF;
+	uint32_t h = (v >> 32) & 0xFFFFFFFF;
+
+	bioAddU32(bio, l);
+	bioAddU32(bio, h);
+}
+
+bool readZipEOCDR(zipEOCDR_t *eocdr, FILE *f)
+{
+	char signature[4];
+
+	fseeko(f, -EOCDR_SIZE, SEEK_END);
+
+	int ret = fread(signature, 1, 4, f);
+
+	if (memcmp(signature, PKZIP_EOCDR_SIGNATURE, 4))
+	{
+		printf("The input file is not a valip zip file - could not find End of Central Directory record\n");
+		return false;
+	}
+
+	eocdr->diskNumber = fileGetU16(f);
+	eocdr->centralDirectoryDiskNumber = fileGetU16(f);
+	eocdr->diskEntries = fileGetU16(f);
+	eocdr->totalEntries = fileGetU16(f);
+	eocdr->centralDirectorySize = fileGetU32(f);
+	eocdr->centralDirectoryOffset = fileGetU32(f);
+	eocdr->commentLen = fileGetU16(f);
+
+	/*if (eocdr->centralDirectoryDiskNumber > 1 || eocdr->diskNumber > 1 ||
+		eocdr->centralDirectoryDiskNumber != eocdr->diskNumber ||
+		eocdr->diskEntries != eocdr->totalEntries)
+	{
+		printf("The input file is a multipart archive - not supported\n");
+		return false;
+	}*/
+
+	if (eocdr->commentLen > 0)
+	{
+		eocdr->comment = calloc(1, eocdr->commentLen + 1);
+		fread(eocdr->comment, 1, eocdr->commentLen, f);
+	}
+	else
+	{
+		eocdr->comment = NULL;
+	}
+
+	return true;
+}
+
+bool readZip64EOCDLocator(zip64EOCDLocator_t *locator, FILE *f)
+{
+	char signature[4];
+
+	fseeko(f, -(EOCDR_SIZE + ZIP64_EOCD_LOCATOR_SIZE), SEEK_END);
+
+	fread(signature, 1, 4, f);
+
+	if (memcmp(signature, PKZIP64_EOCD_LOCATOR_SIGNATURE, 4))
+	{
+		printf("The input file is not a valip zip file - could not find zip64 EOCD locator\n");
+		return false;
+	}
+
+	locator->diskWithEOCD = fileGetU32(f);
+	locator->eocdOffset = fileGetU64(f);
+	locator->totalNumberOfDisks = fileGetU32(f);
+
+	return true;
+}
+
+bool readZip64EOCDR(zip64EOCDR_t *eocdr, FILE *f, uint64_t offset)
+{
+	char signature[4];
+
+	fseeko(f, offset, SEEK_SET);
+
+	fread(signature, 1, 4, f);
+
+	if (memcmp(signature, PKZIP64_EOCDR_SIGNATURE, 4))
+	{
+		printf("The input file is not a valip zip file - could not find zip64 End of Central Directory record\n");
+		return NULL;
+	}
+
+	eocdr->eocdrSize = fileGetU64(f);
+	eocdr->creatorVersion = fileGetU16(f);
+	eocdr->viewerVersion = fileGetU16(f);
+	eocdr->diskNumber = fileGetU32(f);
+	eocdr->diskWithCentralDirectory = fileGetU32(f);
+	eocdr->diskEntries = fileGetU64(f);
+	eocdr->totalEntries = fileGetU64(f);
+	eocdr->centralDirectorySize = fileGetU64(f);
+	eocdr->centralDirectoryOffset = fileGetU64(f);
+	eocdr->commentLen = eocdr->eocdrSize - 44;
+
+	if (eocdr->commentLen > 0)
+	{
+		eocdr->comment = malloc(eocdr->commentLen);

```

Comment:
There is a potential crash here if a malicious archive specifies an excessively large data sector